test/rules/strict-type-predicates/test.ts.lint

```diff
@@ -0,0 +1,154 @@
+declare function get<T>(): T;
+
+// typeof undefined
+{
+    typeof get<boolean>() === "undefined";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    typeof get<void>() === "undefined";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
+
+    typeof get<boolean | undefined>() === "undefined";
+
+    declare const c: any;
+    typeof get<any>() === "undefined";
+
+    // 'undefined' is not assignable to '{}'
+    typeof get<{}>() === "undefined";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+}
+
+// typeof boolean
+{
+    declare const a: boolean;
+    typeof get<boolean>() === "boolean";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
+
+    typeof get<Boolean>() === "boolean";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    typeof get<boolean | undefined>() === "boolean";
+
+    typeof get<{}>() === "boolean";
+}
+
+// typeof number
+{
+    enum E {}
+
+    typeof get<E>() === "number";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
+
+    typeof get<Number>() === "number";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+}
+
+// typeof string
+{
+    typeof get<"abc" | "def">() === "string";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
+
+    typeof get<String>() === "string";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    typeof get<"abc" | undefined>() === "string";
+}
+
+// typeof symbol
+{
+    typeof get<symbol>() === "symbol";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
+
+    typeof get<string>() === "symbol";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    typeof get<symbol | string>() === "symbol";
+}
+
+// typeof function
+{
+    typeof get<() => void>() === "function";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
+
+    typeof get<Function>() === "function";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
+
+
+    typeof get<number>() === "function";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    typeof get<number | (() => void)>() === "function";
+}
+
+// typeof object
+{
+    typeof get<boolean | number | string | symbol | (() => void) | Function>() === "object";
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    typeof get<{}> === "object";
+}
+
+// === null / undefined
+{
+    get<number | undefined>() === null;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    get<number | null>() === null;
+
+    get<number | null>() === undefined;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    get<number | undefined>() === undefined;
+    get<string | undefined>() == null;
```

Comment:
Could this rule also check for non strict comparisons to `null` or `undefined` if only one of them is possible?
Example:
```typescript
get<string|undefined>() == null;
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [use === undefined instead]

get<string|undefined|null>() == null; // still allowed
```